checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInferenceLogic.java

```diff
@@ -0,0 +1,284 @@
+      if (fieldsWithCMV.field.equals(elt)) {
+        return fieldsWithCMV;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Propagates a set of FieldWithCMVs to successors. If the successor is a regular exit point, then
+   * it searches for owning fields by looking at the called methods set stored in FieldWithCMVs and
+   * the must call obligation of each field.
+   *
+   * @param curBlock the current block
+   * @param fieldsWithCMVs the set of FieldWithCMVs for the current block
+   * @param visited block-FieldWithCMVs pairs already on the worklist
+   * @param worklist current worklist
+   */
+  private void propagateRegPaths(
+      BlockWithFieldsFacts curBlock,
+      Set<FieldWithCMVs> fieldsWithCMVs,
+      Set<BlockWithFieldsFacts> visited,
+      Deque<BlockWithFieldsFacts> worklist) {
+
+    Set<Block> successors = curBlock.block.getSuccessors();
+
+    for (Block b : successors) {
+      if (b.getType() == Block.BlockType.SPECIAL_BLOCK) {
+
+        if (((SpecialBlock) b).getSpecialType() == SpecialBlock.SpecialBlockType.EXCEPTIONAL_EXIT) {
+          // Do nothing if the successor block is an exceptional exit point
+          continue;
+        }
+
+        findOwningFields(fieldsWithCMVs, cfgMethod.getMethod());
+      }
+
+      BlockWithFieldsFacts successor = new BlockWithFieldsFacts(b, fieldsWithCMVs);
+      if (visited.add(successor)) {
+        worklist.add(successor);
+      }
+    }
+  }
+
+  /**
+   * Checks all element in the fieldsWithCMVs and adds an {@code @Owning} annotation on a field if
+   * the called methods set of that field in {@link FieldWithCMVs} includes all the methods in
+   * the @MustCall type of the field. It also updates classToFieldToFinalizers with the new detected
+   * finalizer.
+   *
+   * @param fieldsWithCMVs the set of FieldWithCMVs
+   * @param methodTree the method tree
+   */
+  private void findOwningFields(Set<FieldWithCMVs> fieldsWithCMVs, Tree methodTree) {
+    for (FieldWithCMVs fieldsWithCMV : fieldsWithCMVs) {
+      List<String> mcValues = typeFactory.getMustCallValue(fieldsWithCMV.field);
+      if (fieldsWithCMV.calledMethods.containsAll(mcValues)) {
+        WholeProgramInference wpi = typeFactory.getWholeProgramInference();
+        if (wpi != null) {
+          wpi.addFieldDeclarationAnnotation(fieldsWithCMV.field, OWNING);
+          typeFactory.updateClassToFieldToFinalizers(
+              cfgMethod.getClassTree(), fieldsWithCMV.field, TreeUtils.elementFromTree(methodTree));
+        }
+      }
+    }
+  }
+
+  /**
+   * A pair of a {@link Block} and a set of FieldsWithCMV on entry to the block. Each FieldsWithCMV
+   * represents a set of fields with a least upper bound of methods that were called somewhere in
+   * the previous blocks.
+   */
+  private static class BlockWithFieldsFacts {
+
+    /** The block. */
+    public final Block block;
+
+    /** The set of FieldsWithCMV. */
+    public final ImmutableSet<FieldWithCMVs> fieldsWithCMV;
```

Comment:
Could this be a Map from each field to a set of called methods instead?  Representing as a set of pairs seems a bit convoluted and could lead to weird bugs (e.g., since you don't want to allow mutiple entries in the set for the same field)